test/e2e/fishapp/dynamic_stack.go

```diff
@@ -0,0 +1,811 @@
+	Expect(err).NotTo(HaveOccurred())
+	Expect(len(pods.Items)).NotTo(BeZero())
+
+	podsConnectivityCheckResult := make(map[string]map[string]map[string]int)
+	for i := range pods.Items {
+		pod := pods.Items[i].DeepCopy()
+		By(fmt.Sprintf("check pod %s/%s connectivity to services", pod.Namespace, pod.Name), func() {
+			podConnectivityCheckResult := s.checkPodVirtualServiceConnectivity(ctx, f, pod, vsList)
+			podsConnectivityCheckResult[fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)] = podConnectivityCheckResult
+		})
+	}
+	return podsConnectivityCheckResult
+}
+
+func (s *DynamicStack) checkPodVirtualServiceConnectivity(ctx context.Context, f *framework.Framework,
+	pod *corev1.Pod, vsList []*appmeshv1beta1.VirtualService) map[string]map[string]int {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var urls []string
+	for _, vs := range vsList {
+		for _, route := range vs.Spec.Routes {
+			path := route.Http.Match.Prefix
+			urls = append(urls, fmt.Sprintf("http://%s:%d%s", vs.Name, shared.AppContainerPort, path))
+		}
+	}
+
+	readyChan := make(chan struct{})
+	portForwarder, err := k8s.NewPortForwarder(ctx, f.RestCfg, pod, []string{fmt.Sprintf("%d", shared.HttpProxyContainerPort)}, readyChan)
+	Expect(err).NotTo(HaveOccurred())
+
+	errChan := make(chan error)
+	go func(errChan chan error) {
+		errChan <- portForwarder.ForwardPorts()
+	}(errChan)
+
+	proxyUrl, err := url.Parse(fmt.Sprintf("http://localhost:%d", shared.HttpProxyContainerPort))
+	Expect(err).NotTo(HaveOccurred())
+	proxyClient := &http.Client{Transport: &http.Transport{Proxy: http.ProxyURL(proxyUrl)}}
+
+	type urlCheckResult struct {
+		url        string
+		statusCode int
+		body       string
+		err        string
+	}
+
+	urlCheckResultChan := make(chan urlCheckResult)
+	go func() {
+		var wg sync.WaitGroup
+		throttle := time.Tick(connectivityCheckRate)
+		for _, url := range urls {
+			for i := 0; i != s.ConnectivityCheckPerURL; i++ {
+				<-throttle
+				wg.Add(1)
+				go func(url string) {
+					defer wg.Done()
+					<-readyChan
+
+					ret := urlCheckResult{
+						url: url,
+					}
+					resp, err := proxyClient.Get(url)
+					if err != nil {
+						ret.err = err.Error()
+						urlCheckResultChan <- ret
+						return
+					}
+					ret.statusCode = resp.StatusCode
+					body, err := ioutil.ReadAll(resp.Body)
+					if err != nil {
+						ret.err = err.Error()
+						urlCheckResultChan <- ret
+						return
+					}
+					ret.body = string(body)
+					urlCheckResultChan <- ret
+				}(url)
+			}
+		}
+		wg.Wait()
+		close(urlCheckResultChan)
+	}()
+
+	urlRespCount := make(map[string]map[string]int)
+	for ret := range urlCheckResultChan {
+		if _, ok := urlRespCount[ret.url]; !ok {
+			urlRespCount[ret.url] = make(map[string]int)
+		}
+		payload := fmt.Sprintf("code: %v, body: %v, err: %v", ret.statusCode, ret.body, ret.err)
+		urlRespCount[ret.url][payload] += 1
+	}
+	cancel()
+	Expect(<-errChan).To(BeNil())
+	return urlRespCount
```

Comment:
Where are we asserting that a particular percentage of weighted traffic was routed to a specific backend?

Reply:
Added a test based on chi-square under pvalue 0.001 :D